### ISSUE TITLE
Streaming sometimes has no response on repeated same request

### ISSUE THREAD

**Issue by npuichigo** [ajndkr/lanarky#146]
I wrapped lanarky `StreamingResponse` in my own fastapi router, but sometimes repeated same requests have no response.
```
router = LangchainRouter(llm_cache_mode=LLMCacheMode.IN_MEMORY)

@router.post("/conversation")
def chat(request: ChatRequest):
    print(request)
    chain = create_chain(...)
    return StreamingResponse.from_chain(chain, request.message)
```

For example, only the first request has response.
```
(base) ichigo@ichigo:~/LocalCodes/ibot$ curl -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "message": "知道迪迦奥特曼吗？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f66afaa"
}'
当然知道迪迦奥特曼！他是我们很多人的童年英雄，拥有着超人的力量和勇气，保护着地球免受怪兽的侵害。
(base) ichigo@ichigo:~/LocalCodes/ibot$ curl -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "message": "知道迪迦奥特曼吗？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f66afaa"
}'
(base) ichigo@ichigo:~/LocalCodes/ibot$ curl -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "message": "知道迪迦奥特曼吗？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f66afaa"
}'
(base) ichigo@ichigo:~/LocalCodes/ibot$ curl -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "message": "知道迪迦奥特曼吗？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f66afaa"
}'
```

To debug, I add one line before returning `StreamResponse`
```
print(chain(request.message))
```
And actually everytime it has result.
![image](https://github.com/ajndkr/lanarky/assets/11533479/2dcec805-81c6-4092-87d4-a44d8763f024)

**Comment by ajndkr** [comment]
hi @npuichigo!

thanks for reporting this bug. It appears streaming doesn't work when `LLMCacheMode.IN_MEMORY` is enabled. Can you try again without caching and let me know what happens?

**Comment by ajndkr** [comment]
gotcha! so it's definitely a caching bug. without much investigation atm, my best guess is that the request fails because the `on_new_llm_token` callback method is never called when caching is enabled.

I will explore a possible fix for this.

**Comment by npuichigo** [comment]
@ajndkr another strange issue, sometimes I get response simply including 'choice', like:
```
(base) ichigo@ichigo:~$ curl -N -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: text/event-stream'   -H 'Content-Type: application/json'   -d '{
  "message": "你是谁？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f666aaa",
  "preset": "AI_BOT"
}'
我是百川大模型，是由百川智能的工程师们创造的大语言模型，我可以和人类进行自然交流、解答问题、协助创作，帮助大众轻松、普惠的获得世界知识和专业服务。如果 你有任何问题，可以随时向我提问

(base) ichigo@ichigo:~$ curl -N -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: text/event-stream'   -H 'Content-Type: application/json'   -d '{
  "message": "你会什么？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f666aaa",
  "preset": "AI_BOT"
}'
我会回答问题、提供信息、协助创作，帮助大众轻松、普惠的获得世界知识和专业服务。

(base) ichigo@ichigo:~$ curl -N -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: text/event-stream'   -H 'Content-Type: application/json'   -d '{
  "message": "你还会什么？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f666aaa",
  "preset": "AI_BOT"
}'
'choices'

(base) ichigo@ichigo:~$ curl -N -X 'POST'   'http://api.ibyteai.com:16017/api/v1/chat/conversation'   -H 'accept: text/event-stream'   -H 'Content-Type: application/json'   -d '{
  "message": "你还会什么？",
  "conversation_id": "3fa85f64-5717-4562-b3fc-2c963f666aaa",
  "preset": "AI_BOT"
}'
我会回答问题、提供信息、协助创作，帮助大众轻松、普惠的获得世界知识和专业服务。
```

It's strange that the same question may have correct response or just 'choice'

**Comment by ajndkr** [comment]
'choices' as response means there's an unhandled error. I will look into it.

**Comment by npuichigo** [comment]
Like LangServe, maybe do not expose error message to client and use 500 Internal Server Error?

https://github.com/langchain-ai/langserve/blob/7372bcff3a5b548ef82aa0f21260bc02f8145cd8/langserve/server.py#L820-L829

**Comment by ajndkr** [comment]
Sounds good. 

I will need to check if there are any complexities with SSE because one the stream is established, I don't think we can send a 500 error. Please correct me if i'm wrong here.

**Comment by npuichigo** [comment]
It seems [SSE](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#examples) contains data-only message or named event. LangServe return a sse named event with: `event: error, data: {"status_code": 500, "message": "Internal Server Error"}` on error.

**Comment by ajndkr** [comment]
@npuichigo regarding the caching issue. there's nothing to fix in the codebase, you need to pass an additional param when using custom chains. Here's an example:

```
import os

from fastapi import FastAPI
from lanarky import LangchainRouter
from lanarky.responses import StreamingResponse
from langchain.chains import LLMChain
from langchain.chat_models import ChatOpenAI
from langchain.prompts import PromptTemplate
from pydantic import BaseModel

prompt_template = """The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Human: {input}
AI:"""


os.environ["OPENAI_API_KEY"] = "<add-your-key>"


class ChatRequest(BaseModel):
    input: str


app = FastAPI()

router = LangchainRouter(llm_cache_mode=1)


@router.post("/chat")
async def chat(
    request: ChatRequest,
) -> StreamingResponse:
    """Streaming chat endpoint."""
    chain = LLMChain(
        llm=ChatOpenAI(temperature=0, streaming=True, cache=True),
        prompt=PromptTemplate.from_template(prompt_template),
        verbose=True,
    )
    return StreamingResponse.from_chain(
        chain, request.model_dump(), media_type="text/event-stream", callback_kwargs={"output_key": chain.output_key}
    )


app.include_router(router)
```

2 things to note:
- pass `cache=True` when initialising your LLM (in this, it's `ChatOpenAI`)
- pass `callback_kwargs` when initialising `StreamingResponse`

At the moment, I have not implemented "simulated" streaming in case caching is enabled. The endpoint will send the full output in a single event.

Regarding, the 500 error, I will consider it as a different issue and close this one for now. 

Feel free to reopen if the above doesn't work for you.

**Comment by ajndkr** [comment]
@npuichigo i've included the 500 error event in #150. feel free to test it out from the `main` branch. The new changes will be part of v0.8 release (will make the new release in a week or so)
